### PR TITLE
Generated marshaling/unmarshaling of non-native database types

### PR DESCRIPTION
**Pull request by markylaing**
@tomponline @stgraber @masnax Following from the chat in IRC I've written this simple implementation of database marshalling/unmarshalling. This allows us to add any type to the database in a TEXT column, the type should implement `query.MarshalDB` and `query.UnmarshalDB`, and should be added to the database type representation with `marshal=true` in the struct tag for the field.

I have implemented this only for the methods it seemed necessary, and only for simple tables (no reference or association tables).

There are no changes to generation for LXD (save for a quick fix on a comment). I'll submit a draft PR for LXD Cloud shortly to show what this looks like in practice.

### CODE REVIEW

**Comment by markylaing** [comment]
Please see https://github.com/canonical/lxd-cloud/pull/275 for example usage.

File: lxd/db/query/marshal.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package query
+
+import (
+	"fmt"
+)
+
+type DBMarshaler interface {
```
Comment (tomponline) [inline lxd/db/query/marshal.go:7]: As this is inside the db/query package I think we can just call this Marshaler, like the Marsheler in the json package.

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
@markylaing please can you update the docs that @masnax added that documents the different field tags.

File: lxd/db/query/marshal.go (hunk 1 above)
Comment (markylaing) [reply lxd/db/query/marshal.go]: Updated.

**Comment by markylaing** [comment]
> @markylaing please can you update the docs that @masnax added that documents the different field tags.
> 
> https://github.com/lxc/lxd/tree/master/lxd/db/generate

I've added the tag and a description.

**Review by tomponline — APPROVED** [review]

**Pull request merged**